numpy/core/src/multiarray/sequence.c

```diff
@@ -69,22 +69,66 @@ NPY_NO_EXPORT PySequenceMethods array_as_sequence = {
 static int
 array_any_nonzero(PyArrayObject *arr)
 {
-    npy_intp counter;
-    PyArrayIterObject *it;
-    npy_bool anyTRUE = NPY_FALSE;
+    PyArray_NonzeroFunc *nonzero;
+    int ret = NPY_FALSE;
+    int needs_api = 0;
+
+    NpyIter *iter;
+    NpyIter_IterNextFunc *iternext;
+    char **dataptr;
+    npy_intp *strideptr, *innersizeptr;
+    NPY_BEGIN_THREADS_DEF;
+
+    nonzero = PyArray_DESCR(arr)->f->nonzero;
+
+    iter = NpyIter_New(arr, NPY_ITER_READONLY |
+                            NPY_ITER_EXTERNAL_LOOP |
+                            NPY_ITER_REFS_OK,
+                            NPY_KEEPORDER, NPY_NO_CASTING,
+                            NULL);
+    if (iter == NULL) {
+        return ret;
+    }
+    needs_api = NpyIter_IterationNeedsAPI(iter);
 
-    it = (PyArrayIterObject *)PyArray_IterNew((PyObject *)arr);
-    if (it == NULL) {
-        return anyTRUE;
+    /* Get the pointers for inner loop iteration */
+    iternext = NpyIter_GetIterNext(iter, NULL);
+    if (iternext == NULL) {
+        NpyIter_Deallocate(iter);
+        return -1;
     }
-    counter = it->size;
-    while (counter--) {
-        if (PyArray_DESCR(arr)->f->nonzero(it->dataptr, arr)) {
-            anyTRUE = NPY_TRUE;
-            break;
+
+    NPY_BEGIN_THREADS_NDITER(iter);
+
+    dataptr = NpyIter_GetDataPtrArray(iter);
+    strideptr = NpyIter_GetInnerStrideArray(iter);
+    innersizeptr = NpyIter_GetInnerLoopSizePtr(iter);
+
+    /* Iterate over all the elements to count the nonzeros */
+    do {
+        char *data = *dataptr;
+        npy_intp stride = *strideptr;
+        npy_intp count = *innersizeptr;
+
+        while (count--) {
+            if (nonzero(data, arr)) {
+                ret = NPY_TRUE;
+                goto finish;
```

Comment:
Also, I fixed a bug that this line said "break" in the last version you reviewed.